framework/decode/vulkan_buffer_tracker.cpp

```diff
@@ -0,0 +1,105 @@
+** and/or sell copies of the Software, and to permit persons to whom the
+** Software is furnished to do so, subject to the following conditions:
+**
+** The above copyright notice and this permission notice shall be included in
+** all copies or substantial portions of the Software.
+**
+** THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+** IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+** FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+** AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+** LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+** FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+** DEALINGS IN THE SOFTWARE.
+*/
+
+#include "decode/vulkan_buffer_tracker.h"
+
+GFXRECON_BEGIN_NAMESPACE(gfxrecon)
+GFXRECON_BEGIN_NAMESPACE(decode)
+
+VulkanBufferTracker::VulkanBufferTracker(const VulkanObjectInfoTable& object_info_table) :
+    _object_info_table(object_info_table)
+{}
+
+void decode::VulkanBufferTracker::TrackBuffer(const decode::BufferInfo* buffer_info)
+{
+    if (buffer_info != nullptr)
+    {
+        if (buffer_info->capture_address != 0)
+        {
+            _capture_addresses[buffer_info->capture_address] = buffer_info->capture_id;
+        }
+
+        if (buffer_info->replay_address != 0)
+        {
+            _replay_addresses[buffer_info->replay_address] = buffer_info->capture_id;
+        }
+    }
+}
+
+void VulkanBufferTracker::RemoveBuffer(const BufferInfo* buffer_info)
+{
+    if (buffer_info != nullptr)
+    {
+        _capture_addresses.erase(buffer_info->capture_address);
+        _replay_addresses.erase(buffer_info->replay_address);
+    }
+}
+
+const decode::BufferInfo*
+decode::VulkanBufferTracker::GetBufferByReplayDeviceAddress(VkDeviceAddress replay_address) const
+{
+    return GetBufferInfo(replay_address, _replay_addresses);
+}
+
+const decode::BufferInfo*
+decode::VulkanBufferTracker::GetBufferByCaptureDeviceAddress(VkDeviceAddress capture_address) const
+{
+    return GetBufferInfo(capture_address, _capture_addresses);
+}
+
+const BufferInfo* VulkanBufferTracker::GetBufferInfo(VkDeviceAddress                                  device_address,
+                                                     const VulkanBufferTracker::device_address_map_t& address_map) const
+{
+    if (!address_map.empty())
+    {
+        // find first address equal or greater
+        auto address_it = address_map.lower_bound(device_address);
+
+        if (address_it == address_map.end() || address_it->first > device_address)
+        {
+            // not found
+            if (address_it == address_map.begin())
+            {
+                return nullptr;
+            }
+
+            // decrement iterator, now pointing to the first VkDeviceAddress that is lower than device_address
+            address_it--;
+        }
```

Comment:
Wouldn't upper_bound work here as well as it gets the "same value or next lowest"? or does this need to get the next lowest if its the same value?

The logic here is difficult to follow, so I want to be doubly sure I understand it before approving.

Reply:
yes, you're right both `lower_bound` and `upper_bound` would work here without modifying anything else. 

but there is a minor difference: 
lower_bound -> key >= device_address
upper_bound -> key > device_address

`lower_bound` 'can' yield a key **equal** to device_address, allowing to skip the following if-statement,
while upper_bound will always be greater than device_address, which will be followed by the iterator-decrement.

so, with both lower_bound and upper_bound we are in the exact same situation after the if-statement:
-> `// found_address is lower or equal to device_address`

Reply:
side-comment:
lower_bound/upper_bound make slightly more sense for a `std::multimap`,
there defining a sub-range with equal keys.
so in cases like this, for regular `std::map`, they appear 'almost' identical in practice